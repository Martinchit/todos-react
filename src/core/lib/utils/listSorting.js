export const listSortingByCompletion = (key) => (a, b) => a[key] - b[key];
