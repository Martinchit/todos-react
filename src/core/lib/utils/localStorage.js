export const retrieveToDos = () => {
  const result = localStorage.getItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY);
  return result ? JSON.parse(result) : null;
};

export const updateToDos = (toDos) => {
  const stringified = JSON.stringify(toDos);
  localStorage.setItem(process.env.REACT_APP_TODO_LOCAL_STORAGE_KEY, stringified);
};
