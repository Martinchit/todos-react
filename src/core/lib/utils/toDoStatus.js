export const TO_DO_STATUS = {
  ALL: 'all',
  COMPLETED: 'completed',
  ACTIVE: 'active',
};
