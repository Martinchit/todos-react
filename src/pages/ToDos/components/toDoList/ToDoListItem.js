import React from 'react';
import Switch from 'react-switch';
import PropTypes from 'prop-types';
import {
  ToDoItemContainer,
  ToDoLabel,
  ToDoEditInput,
  ItemDeleteButton,
} from './styled';
import { useCloneToDo } from '../../../../core/lib/hooks/useCloneToDo';
import { ENTER_KEY } from '../../../../core/lib/utils/inputEventKey';

export const ToDoListItem = ({ toDo, onToDoUpdate, onToDoDelete }) => {
  const [isEditMode, setIsEditMode] = React.useState(false);
  const [toDoClone, setToDoClone] = useCloneToDo(toDo);

  const onToDoCloneUpdate = React.useCallback(
    ({ target }) => {
      setToDoClone({ ...toDoClone, title: target.value });
    },
    [toDoClone]
  );

  const toggleEditMode = React.useCallback(() => {
    setIsEditMode(!isEditMode);
  }, [isEditMode, setIsEditMode]);

  const onKeyUp = React.useCallback(
    ({ code }) => {
      if (code === ENTER_KEY) {
        onToDoUpdate(toDoClone);
        toggleEditMode();
      }
    },
    [onToDoUpdate, toDoClone]
  );

  const onInputBlur = React.useCallback(() => {
    onToDoUpdate(toDoClone);
    toggleEditMode();
  }, [onToDoUpdate, toDoClone]);

  const onToDoStatusUpdate = React.useCallback(
    (completed) => {
      onToDoUpdate({ ...toDoClone, completed });
    },
    [onToDoUpdate, toDoClone]
  );

  return (
    <ToDoItemContainer
      width="96%"
      justifyContent="space-between"
      padding="5px 2.5%"
      height="40px"
    >
      <Switch
        checked={toDo.completed}
        onChange={onToDoStatusUpdate}
        width={40}
        height={20}
      />
      {!isEditMode ? (
        <ToDoLabel onDoubleClick={toggleEditMode}>{toDoClone.title}</ToDoLabel>
      ) : (
        <ToDoEditInput
          type="text"
          value={toDoClone.title}
          onChange={onToDoCloneUpdate}
          onKeyUp={onKeyUp}
          onBlur={onInputBlur}
          autoFocus
        />
      )}
      <ItemDeleteButton onClick={() => onToDoDelete(toDo.id)} />
    </ToDoItemContainer>
  );
};

ToDoListItem.propTypes = {
  toDo: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.string,
    completed: PropTypes.bool,
  }),
  onToDoUpdate: PropTypes.func,
  onToDoDelete: PropTypes.func,
};

ToDoListItem.defaultProps = {
  toDo: {},
  onToDoUpdate: null,
  onToDoDelete: null,
};
