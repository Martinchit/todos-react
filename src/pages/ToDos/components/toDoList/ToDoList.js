import React from 'react';
import PropTypes from 'prop-types';
import { ToDoListItem } from './ToDoListItem';
import { Box } from '../../../../core/components/atoms/Box';
import { TO_DO_STATUS } from '../../../../core/lib/utils/toDoStatus';

export const ToDoList = ({
  toDos,
  onToDoUpdate,
  onToDoDelete,
  currentFilterOption,
}) => {
  return (
    <Box flexDirection="column">
      {toDos
        .filter((tD) =>
          currentFilterOption === TO_DO_STATUS.ALL
            ? tD
            : currentFilterOption === TO_DO_STATUS.COMPLETED
            ? tD.completed
            : !tD.completed
        )
        .map((toDo) => (
          <ToDoListItem
            toDo={toDo}
            key={toDo.id}
            onToDoUpdate={onToDoUpdate}
            onToDoDelete={onToDoDelete}
          />
        ))}
    </Box>
  );
};

ToDoList.propTypes = {
  toDos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      completed: PropTypes.bool,
    })
  ),
  onToDoUpdate: PropTypes.func,
  onToDoDelete: PropTypes.func,
  currentFilterOption: PropTypes.oneOf(Object.values(TO_DO_STATUS)),
};

ToDoList.defaultProps = {
  toDos: [],
  onToDoUpdate: null,
  onToDoDelete: null,
  currentFilterOption: TO_DO_STATUS.ALL,
};
