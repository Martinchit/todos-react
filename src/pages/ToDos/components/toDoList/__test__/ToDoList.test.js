import React from 'react';
import { shallow } from 'enzyme'; 
import 'jest-styled-components';

import { ToDoList } from '../ToDoList';
import { ToDoListItem } from '../ToDoListItem';
import { TO_DO_STATUS } from '../../../../../core/lib/utils/toDoStatus';

describe('ToDoList', () => {
  const onToDoUpdate = jest.fn();
  const onToDoDelete = jest.fn();
  const toDos = [
    {
      id: '1',
      title: 'delectus aut autem',
      completed: true
    },
    {
      id: '2',
      title: 'delectus aut autem',
      completed: false
    },
    {
      id: '3',
      title: 'delectus aut autem',
      completed: true
    }
  ]
  it('renders correctly', () => {
    const wrapper = shallow(<ToDoList toDos={toDos} onToDoUpdate={onToDoUpdate} onToDoDelete={onToDoDelete } />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(ToDoListItem).length).toBe(3);
  })

  it('renders correct number of completed task', () => {
    const wrapper = shallow(<ToDoList toDos={toDos} currentFilterOption={TO_DO_STATUS.COMPLETED} onToDoUpdate={onToDoUpdate} onToDoDelete={onToDoDelete } />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(ToDoListItem).length).toBe(2);
  })

  it('renders correctly', () => {
    const wrapper = shallow(<ToDoList toDos={toDos} currentFilterOption={TO_DO_STATUS.ACTIVE} onToDoUpdate={onToDoUpdate} onToDoDelete={onToDoDelete } />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(ToDoListItem).length).toBe(1);
  })

  it('handle onToDoDelete', () => {
    const wrapper = shallow(<ToDoList toDos={toDos} onToDoUpdate={onToDoUpdate} onToDoDelete={onToDoDelete } />)
    const firstToDoItem = wrapper.find(ToDoListItem).at(0);

    firstToDoItem.props().onToDoDelete();
    expect(onToDoDelete).toHaveBeenCalled();

    firstToDoItem.props().onToDoUpdate();
    expect(onToDoUpdate).toHaveBeenCalled();
  })
})
