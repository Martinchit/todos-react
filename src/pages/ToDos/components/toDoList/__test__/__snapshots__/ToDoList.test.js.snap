// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList renders correct number of completed task 1`] = `ShallowWrapper {}`;

exports[`ToDoList renders correctly 1`] = `ShallowWrapper {}`;

exports[`ToDoList renders correctly 2`] = `ShallowWrapper {}`;
